src/audio_pixel.h

```diff
@@ -37,45 +44,76 @@ class audio_pixel_t {
     void to_json(json& j) const;
     void from_json(const json& j);
 
+    const int get_mono_pixel_count() {
+        if (m_channel_pixels.size() > 0)
+            return m_channel_pixels[0].size();
+        return 0;
+    }
+    const int get_stero_pixel_count() {
+        if (m_channel_pixels.size() > 0)
+            return m_channel_pixels[0].size() * m_channel_pixels.size();
+        return 0;
+    }
+    const std::vector<std::vector<audio_pixel_t>>& get_pixels() const { return m_channel_pixels; }
+    const std::vector<std::vector<audio_pixel_t>>& get_pixels(double t0, double t1) const { /* TODO */}
+    double get_pps() { return m_pix_per_s; }
+
+
+    void update(std::vector<double>& sample_buffer, int num_channels, int sample_rate);
+
+
+
 private: 
-    MediaItem_Take *m_take {nullptr}; // parent take
+    std::vector<std::vector<audio_pixel_t>> m_channel_pixels; // vector of audio_pixels for each channel 
+    std::string m_guid; // unique-id mapping back to a MediaItemTake, fill w/ GetSetMediaItemTakeInfo_String 
     double m_pix_per_s {1.0}; // pixels per second
-    std::vector<audio_pixel_t> m_pixels;  // stores pixel data
-    safe_audio_accessor_t m_accessor; // to get samples from the MediaItem_Take
-    std::string m_guid; // unique id that maps back to a MediaItemTake
-                        // use GetSetMediaItemTakeInfo_String to fill it up
+
 };
 
 // hold audio_pixel_block_t at different resolutions
 // and is able to interpolate between them to 
 // get audio pixels at any resolution inbetween
 class audio_pixel_mipmap_t {
-    audio_pixel_mipmap_t(MediaItem_Take* take);
+    audio_pixel_mipmap_t(MediaItem_Take* take)
+        :m_accessor(safe_audio_accessor_t(take)),
+        m_take(take)
+    {
+        if (!m_accessor.is_valid()) { std::cerr << "Invalid audio accessor used for audio block." << std::endl; }
+    };
 
-    void update(); // update all blocks
     bool flush() const; // flush all blocks to disk
 
     // mm, how do we pass these pixels fast enough?
-    std::vector<audio_pixel_t> get_pixels(double t0, double t1, double pix_per_s) const;
+    audio_pixel_block_t get_pixels(double t0, double t1, double pix_per_s);
+
+    void update();
 
 private:
+    double closet_val(double val1, double val2, double target);
+    double get_nearest_block_pps(double pix_per_s);
+    double get_nearest_pps(double pix_per_s);
+    audio_pixel_block_t interpolate_block(double t0, double t1, double new_pps, double nearest_pps);
+    double linear_interp(double x, double x1, double x2, double y1, double y2);
+    
     MediaItemTake* m_take {nullptr};
-    std::map<int, audio_pixel_block_t> m_blocks;
-};
\ No newline at end of file
+    safe_audio_accessor_t m_accessor; // to get samples from the MediaItem_Take
+    std::map<double, audio_pixel_block_t> m_blocks;  
+    std::vector<double> m_block_pps; // sorted list of the blocks pps, TODO fill this up after construction
+};
+
+struct cmp_blocks_pps {
```

Comment:
is this comparator needed? i feel like this is the default for the standard library

Reply:
true, `std::greater<double>`